### PR TITLE
Added 't to contractions Map.

### PR DESCRIPTION
**Pull request by wholder**
Added to support contractions such as isn't, wasn't, can't, and so on.

### CODE REVIEW

File: marytts-languages/marytts-lang-en/src/main/java/marytts/language/en/Preprocess.java [hunk 1]
```diff
@@ -153,6 +153,7 @@
 		contractions.put("'d", new String[] { "d" });
 		contractions.put("'m", new String[] { "m" });
 		contractions.put("'re", new String[] { "r" });
+		contractions.put("'t", new String[] {"t"});
```
Comment (psibre) [inline marytts-languages/marytts-lang-en/src/main/java/marytts/language/en/Preprocess.java:156]: >  isn't, wasn't, can't, and so on.

All of the examples [you mention](https://github.com/marytts/marytts/pull/1003#issue-428859932) seem to also include an *n* before the apostrophe.

```suggestion
		contractions.put("n't", new String[] {"t"});
```
Wouldn't this work and capture these examples (and others)?

**Comment by psibre** [comment]
Looks like the contractions are mainly in this part of the spaghetti code:

https://github.com/marytts/marytts/blob/e70e56798806fd342db99bd1a1e837cf857790dd/marytts-languages/marytts-lang-en/src/main/java/marytts/language/en/Preprocess.java#L297-L315

But there is also this bit:

https://github.com/marytts/marytts/blob/e70e56798806fd342db99bd1a1e837cf857790dd/marytts-languages/marytts-lang-en/src/main/java/marytts/language/en/Preprocess.java#L404-L412

and it looks like

https://github.com/marytts/marytts/blob/e70e56798806fd342db99bd1a1e837cf857790dd/marytts-languages/marytts-lang-en/src/main/java/marytts/language/en/Preprocess.java#L567-L572

has no test coverage.

It would be great to add some (parameterized) unit tests in [PreprocessTest](https://github.com/marytts/marytts/blob/d8d5112fb9effd7e009b7ccdd36f68217b6af4dd/marytts-languages/marytts-lang-en/src/test/java/marytts/language/en/PreprocessTest.java) *before* we refactor, so that we can be confident we're not introducing regressions.

Perhaps we'll also discover a lingering bug or two...

**Comment by wholder** [comment]
As currently coded, this block of code will never be executed:

```
 // if tokens are an expanded contraction 
 if (splitContraction && newTokens.length == 2) { 
    if (newTokens[0].substring(newTokens[0].length() - 1).matches("[cfkpt]") 
      && contractions.get(newTokens[i]).length > 1) { 
        t.setAttribute("ph", contractions.get(newTokens[i])[1]); 
    } else { 
      t.setAttribute("ph", contractions.get(newTokens[i])[0]); 
    } 
 } 
```

so the function `String splitContraction(String contraction) `will never be called.  This because this block of coded has been commented out:

```
  // FIXME: we do not want to have to phonological word => for now we do not split !
  / // if not in lexicon and we have a contraction expansion then split into two tokens
  // else
  // {
  // splitContraction = true;
  // MaryDomUtils.setTokenText(t, splitContraction(tokenText));
  // }

```

Wayne

File: marytts-languages/marytts-lang-en/src/main/java/marytts/language/en/Preprocess.java (hunk 1 above)
Comment (wholder) [reply marytts-languages/marytts-lang-en/src/main/java/marytts/language/en/Preprocess.java:156]: I don't think it makes a difference.  However, as as my further comments below suggest, I think the code for handling contractions is currently incomplete.

**Comment by wholder** [comment]
I stumbled upon what might be an additional issue with the `'t` contraction.  In the `postlexicalRules()` method `in en.PronunciationModel.java` there appears to be code to perform some kind of additional processing on `'s`, `'ve`, `'ll` and `'d`.  I'm not exactly sure what this code does, but it seemed related to I'd thought I'd mentioned it in case some additional work is needed here as a consequence of adding `'t` to the contractions table in Preprocess.java.  I'm keeping my changes as a draft until I need more certain that my changes are going to be OK.

**Comment by psibre** [comment]
Thanks very much for the detailed analysis and help!!

For the record, as I recall, the class (and tests) were written by a student some years ago, but to be fair, the code that it replaced was an absolute horror, e.g.,

https://github.com/marytts/marytts/blob/4ee46c9345e4eeb9deaa2a2911a5b3dadb923eff/marytts-languages/marytts-lang-en/src/main/java/marytts/language/en/TokenToWords.java

That previous version of the MaryTTS English NLP code was basically just a wrapper around some custom, repackaged version of [FreeTTS](https://freetts.sourceforge.io/), Sun's Java Port of [CMU Flite](http://www.festvox.org/flite/), itself a streamlined port of [Festival](http://www.cstr.ed.ac.uk/projects/festival/).

We were so happy to wean ourselves off of that mostly unmaintained, third-party baggage at the time that anything that worked reasonably well, was good enough. Unfortunately, following that, there were limited resources to continue to refactor the English NLP code, and here we are.

**Comment by wholder** [comment]
I understand.  I've worked on many projects that have gone through similar growing pains, so I'm happy to help out where I can.

I think I've done all I can to `Preprocess` without getting into issues I don't understand all that well, such as trying to complete the incomplete code for contractions and quotes. It seems to pass all the tests, including a few new ones I've added.  I do think it would be better to rework the test code so that everything is tested by calling either `process()` or `expand()`, but I don't yet understand the Document-based data structures used to pass data in and out well enough to take on a task like that.  So, should I go ahead and click the "Ready for review" button?

Also, there are number of small, but potentially dangerous issues scattered around in the code that I think should be addressed.  Things like using "`==`" instead of equals to compare objects, such as "`if (distanceType == "Mahalanobis"`)" in `DTW.java` (just one example of many.)  I can easily fix these, but this might involve a pull request covering some 20 files for just this one issue.  Is that OK to do, or do you prefer a more granular, one file at a time approach?

In addition, there are also a number of performance-related issues, such as String concatenation in loops that should be converted to use `StringBuilder()`, places where manual array copies are done rather than using `System.arraycopy()`, as well as quite a bit of code that could be modernized to use the `Collections` class more efficiently.  But, again, the most efficient way to fix these things would involve changing quite a few files.

Finally, can I assume that it is safe to use language contracts from Java 8, or earlier?  If so, there are lot of areas in the code that could be streamlined a bit by using some of these newer constructs.

Wayne

**Comment by psibre** [comment]
Thanks very much for your willingness to help, and for the help itself!

Feel free to mark this draft PR as "Ready for review", although I can't say when we'll have time to deal with it. @seblemaguer?

As I mentioned elsewhere, many parts of the code are quite old and violate many coding standards (and probably contain numerous bugs). So if any kind of static code analysis yields low-hanging fruit on what to improve for efficiency or robustness, then please don't hold back! I'm all for new PRs that consistently fix a random problem across entire sets of files, and my only woe is that the test coverage is so low that we can't be sure if it will break anything. But at the end of the day, we do have various integration tests in place, so hopefully one of those will flag a regression... =)

Regarding Java compatibility: yes, our source and target is Java 8.

**Comment by seblemaguer** [comment]
@wholder when you are ready, mark it as ready to review. I will have a look in the following couple of days.

Thanks a lot

**Comment by wholder** [comment]
> @wholder when you are ready, mark it as ready to review. I will have a look in the following couple of days.
> 
> Thanks a lot

I found another regression I introduced and found that it, and several other possible ways to introduce regressions were not covered by the test code.  So, I came up with a way to rewrite the text code to run the tests though the expand() function rather than call the methods that expand() uses.  Still far from perfect, but I think it does a better job of testing the code and it's now easier to add new tests.

I'm going to take a break and thenlook this over again before I check it in, just in case I missed anything else.

Wayne

**Comment by seblemaguer** [comment]
Thanks a lot for the work. I have a look at it tomorrow.

For the non-fixed part, we can see that later :)

**Comment by seblemaguer** [comment]
We discussed with @psibre. It is fine on our side but the spacing is a problem. He will face the formatting

**Comment by wholder** [comment]
Thanks.  For future reference, do you have any coding style guidelines?  At psibre's suggestion, I'm currently in the process of attempting a fairly extensive rewrite to try and address some of the issue I've mentioned recently, as well as removing unnecessary code, such as:
```
  covMatrix[0] = new double[dimension];
  //Arrays.fill(covMatrix[0], 0.0);
```
as the JVM [zero (and false and null) fills](https://www.tutorialspoint.com/what-are-the-default-array-values-in-java) all single dimension arrays by default.  I realize a lot of this code is old and was transcoded from C, so I can understand where this comes from.  So, I'm also trying to recode any C-like code into Java, as well rework code to take better advantage of the Java class libraries.

I know that coding style can be a contentious issue but, if possible, I highly recommend always using {} to establish scope as I find that relying on indenting for scope can hide, or introduce errors when editing.  (See [section 7.4 in "Coding Conevtions for Java"](https://www.oracle.com/technetwork/java/javase/documentation/codeconventions-142311.html#449).)

I also tend to avoid adding unnecessary blank lines inside methods, as this make it possible to see more of the method's code at one time.  Likewise, I think it's best to avoid adding essentially blank Javadoc comments to methods (ones that simply restate the method signature) as these take up screen space while adding nothing of value.

In addition, as Java supports declaring local variables anywhere in a method, I think it's good practice to combine declaration and assignment, when possible, and keep the declaration as close as possible to where it is used.  Again, I know a lot of this in MaryTTS comes from using C code as a template, but it also often leads to redundant initializations and unused assignments as well as makes code harder to read, in some cases.

Wayne

**Comment by psibre** [comment]
Thanks very much for your thoughts!

Indeed, before the MaryTTS codebase was migrated from Subversion to GitHub in 2011, there was no pull request review process at all, and various people (with "varying" degrees of Java experience) had worked on the project or contributed by committing to `trunk` with little or no coding style guidelines.

By 2010, it had gotten so out of hand that @marc1s made a presentation for error handling:

https://github.com/marytts/marytts/commits/master/doc/ErrorHandling.odp

Unfortunately, as the obsolete [TRAC wiki](https://mary.opendfki.de/trac/wiki) still says today,
> Guidelines for code development:
>
> * How to work with MARY source code in Eclipse
> * Code quality guidelines:
>     * Error Handling in MARY TTS
>     * more to follow...

Sometime later (in b9a3e4fc28980d4c3081d9e93860234d2afb321d) we actually committed the code formatting definition for Eclipse (Luna at the time), but this became obsolete when we switched to developing in IntelliJ IDEA.

Sometime after that, we even dabbled in using [astyle](http://astyle.sourceforge.net/) as an IDE-independent code style checker/formatter, but using this again required quite a bit of extra overhead and was never widely adopted.

Of course, this was all strange and exotic to various undergrads and other people contributing to the code over the years, who had never heard of [Clean Code](https://cleancoders.com/) etc...

But I've found that code formatting and static analysis built right into IntelliJ IDEA works better than any half-assed, manually maintained solution. And the defaults are just fine.

## TL;DR Just use IntelliJ IDEA

That's our code style. =)

**Comment by wholder** [comment]
> TL;DR Just use IntelliJ IDEA

So, it's cool if run Code->Reformat Code on all the Java source files?

Wayne

**Comment by psibre** [comment]
> So, it's cool if run Code->Reformat Code on all the Java source files?

Sure! Ideally as a separate commit before a class is changed, so that the commits in a PR have a visible effect in Git blame. =)

**Comment by seblemaguer** [comment]
@wholder if that is fine for you, I close this PR and merge the PR #1005?

**Comment by wholder** [comment]
Yes.  Thanks.
